torch/csrc/jit/operator.cpp

```diff
@@ -457,10 +457,20 @@ bool Operator::matches(const Node* node) const {
     const MatchTypeReturn matched_type =
         matchTypeVariables(formals[i].type(), actuals[i]->type(), type_env);
     if (!matched_type.type) {
+      if (actuals[i]->type() == NoneType::get() &&
+          formals[i].type()->kind() == TypeKind::OptionalType) {
+        // when looking for a match, None is actually OK here
+        continue;
+      }
```

Comment:
Uh how come do we need this special case here, and how did we survive without it before? We should be able to unify a `NoneType` with an `OptionalType`, right?

Reply:
Well, NoneType isn't in a Subtype relation with OptionalType, so we won't find signatures. Before we had an OptionalType going in.

Reply:
Hmm ok, I guess we could change it? Those conditionals are worrying for me. I'm not really sure how our type hierarchy looks right now, and if it wouldn't mess up some things. cc @zdevito @eellison @wanchaol

Reply:
I added it because looked like it's intentional: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L221

Reply:
I think this is fine as well but maybe @zdevito should weigh in

Reply:
Our `None` story is extremely confusing to be honest. The comment you linked says this as well
> instead NoneType is convertable in schema matching to any Optional[T]

which seems to imply that what you're doing in here should have been already present in the code?

I'd really like to hear what @zdevito thinks about this because I haven't been fully following the `None` changes, but it looks very weird to me.

Reply:
This code cannot be modified in this way. NoneType values should not appear in the graph _anywhere_ after the compiler has run. All None's need to be specialized to a particular type otherwise logic related to unwrapping options will no longer work.  Allow None here will cause type variable matching to fail.